### PR TITLE
Add immutable feature blocks

### PR DESCRIPTION
**Pull request by thibault-martinez**
## Change checklist

Tick the boxes that are relevant to your changes, and delete any items that are not.

- [x] I have followed the contribution guidelines for this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked that new and existing unit tests pass locally with my changes
- [x] I have updated the CHANGELOG.md, if my changes are significant enough

### CODE REVIEW

File: bee-message/src/output/alias.rs [hunk 1]
```diff
@@ -171,6 +195,8 @@ pub struct AliasOutput {
     unlock_conditions: UnlockConditions,
     //
     feature_blocks: FeatureBlocks,
+    //
```
Comment (Thoralf-M) [inline bee-message/src/output/alias.rs:198]: What's up with the blank comments? Are they supposed to be done later in another PR?

File: bee-message/src/output/alias.rs [hunk 2]
```diff
@@ -116,6 +118,23 @@ impl AliasOutputBuilder {
         self
     }
 
+    ///
```
Comment (Alex6323) [inline bee-message/src/output/alias.rs:121]: docs

File: bee-message/src/output/alias.rs (hunk 1 above)
Comment (thibault-martinez) [reply bee-message/src/output/alias.rs:198]: Yes, once the TIPs are stable enough (they probably are now), I'll do one big PR with all the docs of bee-message

File: bee-message/src/output/alias.rs [hunk 3]
```diff
@@ -116,6 +118,23 @@ impl AliasOutputBuilder {
...
+    #[inline(always)]
```
Comment (Alex6323) [inline bee-message/src/output/alias.rs:122]: are we sure to require this?

File: bee-message/src/output/alias.rs [hunk 4]
```diff
@@ -116,6 +118,23 @@ impl AliasOutputBuilder {
...
+    pub fn add_immutable_feature_block(mut self, immutable_feature_block: FeatureBlock) -> Self {
+        self.immutable_feature_blocks.push(immutable_feature_block);
+        self
+    }
+
+    ///
```
Comment (Alex6323) [inline bee-message/src/output/alias.rs:128]: docs

File: bee-message/src/output/alias.rs [hunk 5]
```diff
@@ -116,6 +118,23 @@ impl AliasOutputBuilder {
...
+    #[inline(always)]
```
Comment (Alex6323) [inline bee-message/src/output/alias.rs:129]: sure?

File: bee-message/src/output/alias.rs [hunk 6]
```diff
@@ -116,6 +118,23 @@ impl AliasOutputBuilder {
...
+    pub fn with_immutable_feature_blocks(
+        mut self,
+        immutable_feature_blocks: impl IntoIterator<Item = FeatureBlock>,
+    ) -> Self {
+        self.immutable_feature_blocks = immutable_feature_blocks.into_iter().collect();
+        self
+    }
+
     ///
```
Comment (Alex6323) [inline bee-message/src/output/alias.rs:138]: docs

File: bee-api/bee-rest-api/src/types/dtos.rs [hunk 7]
```diff
@@ -1594,9 +1594,6 @@ pub struct NftOutputDto {
     // Unique identifier of the NFT.
     #[serde(rename = "nftId")]
     pub nft_id: NftIdDto,
-    // Binary metadata attached immutably to the NFT.
-    #[serde(rename = "immutableData")]
-    pub immutable_data: String,
```
Comment (Thoralf-M) [inline bee-api/bee-rest-api/src/types/dtos.rs:1599]:
```Rust
    #[serde(rename = "immutableFeatureBlocks")]
    pub immutable_feature_blocks: Vec<FeatureBlockDto>,
```
is missing

File: bee-message/src/output/foundry.rs [hunk 8]
```diff
@@ -119,6 +121,23 @@ impl FoundryOutputBuilder {
         self
     }
 
+    ///
+    #[inline(always)]
```
Comment (Alex6323) [inline bee-message/src/output/foundry.rs:125]: docs/inline

File: bee-message/src/output/foundry.rs [hunk 9]
```diff
@@ -119,6 +121,23 @@ impl FoundryOutputBuilder {
...
+    pub fn add_immutable_feature_block(mut self, immutable_feature_block: FeatureBlock) -> Self {
+        self.immutable_feature_blocks.push(immutable_feature_block);
+        self
+    }
+
+    ///
+    #[inline(always)]
```
Comment (Alex6323) [inline bee-message/src/output/foundry.rs:132]: docs/inline

File: bee-message/src/output/foundry.rs [hunk 10]
```diff
@@ -172,6 +200,8 @@ impl FoundryOutput {
     const ALLOWED_UNLOCK_CONDITIONS: UnlockConditionFlags = UnlockConditionFlags::IMMUTABLE_ALIAS_ADDRESS;
     /// The set of allowed [`FeatureBlock`]s for an [`FoundryOutput`].
```
Comment (Alex6323) [inline bee-message/src/output/foundry.rs]: an -> a

File: bee-message/src/output/foundry.rs [hunk 11]
```diff
@@ -172,6 +200,8 @@ impl FoundryOutput {
...
     pub const ALLOWED_FEATURE_BLOCKS: FeatureBlockFlags = FeatureBlockFlags::METADATA;
+    /// The set of allowed immutable [`FeatureBlock`]s for an [`FoundryOutput`].
```
Comment (Alex6323) [inline bee-message/src/output/foundry.rs]: docs: an -> a

File: bee-message/src/output/foundry.rs [hunk 12]
```diff
@@ -295,6 +325,12 @@ impl FoundryOutput {
     pub fn feature_blocks(&self) -> &[FeatureBlock] {
         &self.feature_blocks
     }
+
+    ///
+    #[inline(always)]
```
Comment (Alex6323) [inline bee-message/src/output/foundry.rs:330]: docs/inline

File: bee-api/bee-rest-api/src/types/dtos.rs (hunk 7 above)
Comment (Thoralf-M) [reply bee-api/bee-rest-api/src/types/dtos.rs:1599]: Also for foundry and alias output dtos

File: bee-message/src/output/nft.rs [hunk 13]
```diff
@@ -31,23 +29,21 @@ pub struct NftOutputBuilder {
...
     feature_blocks: Vec<FeatureBlock>,
+    immutable_feature_blocks: Vec<FeatureBlock>,
 }
 
 impl NftOutputBuilder {
     ///
-    pub fn new(amount: u64, nft_id: NftId, immutable_metadata: Vec<u8>) -> Result<NftOutputBuilder, Error> {
-        verify_immutable_metadata_length(immutable_metadata.len())?;
-
+    pub fn new(amount: u64, nft_id: NftId) -> Result<NftOutputBuilder, Error> {
```
Comment (Alex6323) [inline bee-message/src/output/nft.rs:39]: docs

File: bee-message/src/output/nft.rs [hunk 14]
```diff
@@ -93,6 +89,23 @@ impl NftOutputBuilder {
         self
     }
 
+    ///
+    #[inline(always)]
```
Comment (Alex6323) [inline bee-message/src/output/nft.rs:93]: docs/inline

File: bee-message/src/output/nft.rs [hunk 15]
```diff
@@ -93,6 +89,23 @@ impl NftOutputBuilder {
...
+    pub fn add_immutable_feature_block(mut self, immutable_feature_block: FeatureBlock) -> Self {
+        self.immutable_feature_blocks.push(immutable_feature_block);
+        self
+    }
+
+    ///
+    #[inline(always)]
```
Comment (Alex6323) [inline bee-message/src/output/nft.rs:100]: docs/inline

File: bee-message/src/output/nft.rs [hunk 16]
```diff
@@ -93,6 +89,23 @@ impl NftOutputBuilder {
...
+    pub fn with_immutable_feature_blocks(
+        mut self,
+        immutable_feature_blocks: impl IntoIterator<Item = FeatureBlock>,
+    ) -> Self {
+        self.immutable_feature_blocks = immutable_feature_blocks.into_iter().collect();
+        self
+    }
+
     ///
```
Comment (Alex6323) [inline bee-message/src/output/nft.rs:109]: docs

File: bee-message/src/output/nft.rs [hunk 17]
```diff
@@ -130,17 +141,14 @@ pub struct NftOutput {
...
     nft_id: NftId,
-    // Binary metadata attached immutably to the NFT.
-    immutable_metadata: BoxedSlicePrefix<u8, ImmutableMetadataLength>,
     unlock_conditions: UnlockConditions,
     feature_blocks: FeatureBlocks,
+    immutable_feature_blocks: FeatureBlocks,
 }
 
 impl NftOutput {
     /// The [`Output`](crate::output::Output) kind of a [`NftOutput`].
```
Comment (Alex6323) [inline bee-message/src/output/nft.rs]: a -> an

File: bee-message/src/output/nft.rs [hunk 18]
```diff
@@ -211,6 +222,12 @@ impl NftOutput {
     pub fn feature_blocks(&self) -> &[FeatureBlock] {
         &self.feature_blocks
     }
+
+    ///
+    #[inline(always)]
```
Comment (Alex6323) [inline bee-message/src/output/nft.rs:227]: docs/inline?

File: bee-message/src/output/alias.rs (hunk 2 above)
Comment (thibault-martinez) [reply bee-message/src/output/alias.rs:121]: As said, there will be a documentation PR

File: bee-api/bee-rest-api/src/types/dtos.rs (hunk 7 above)
Comment (thibault-martinez) [reply bee-api/bee-rest-api/src/types/dtos.rs:1599]: Thanks

File: bee-message/src/output/alias.rs (hunk 3 above)
Comment (thibault-martinez) [reply bee-message/src/output/alias.rs:122]: I am keeping them to keep consistent for now. Can't do harm on such a little functions IMO. Feel free to bring the discussion on Discord with @pvdrz if you have doubts/questions.

File: bee-api/bee-rest-api/src/types/dtos.rs (hunk 7 above)
Comment (thibault-martinez) [reply bee-api/bee-rest-api/src/types/dtos.rs:1599]: Done

**Review by Thoralf-M — APPROVED** [review]

**Review by grtlr — APPROVED** [review]

**Review by Adam-Gleave — APPROVED** [review]

**Pull request merged**